### PR TITLE
SetIfAbsent

### LINKED ISSUE #75: Set() to return a value indicating whether the key already existed
The `Set` method is implemented as blind upsert and does not return any indication of what value, if any, previously existed for the key in question. Callers that wish to leave any existing value intact, or save or report it elsewhere after it is overwritten, must first perform a `Get` or `Has` in order to establish if one exists. `Set` could instead return an indication of whether the key previously existed.

Some options:
- `Set` returns a boolean if the key existed
- `Set` accepts a nillable `out` parameter like `Get`, which receives any previous value (combined with the boolean)
- `Set` or a variant of it accepts a callback receiving the bytes previously stored for the key
- [Edit] A new `SetIfAbsent` writes a key only if not previously present, and returns a boolean indicating whether it did so

The concrete motivating case is safety checks in the Miner actor. It contains a HAMT of pre-committed sectors by sector number. The `PreCommitSector` method wants to abort if a pre-commit collides with a sector number already pre-committed or proven. If it does, the method will abort and roll back all state changes. This method is expensive and we wish to remove all unnecessary state access. The usual case is no collision.

A single boolean return value is sufficient here: even if the HAMT is internally modified by the `Set`, the modification will be thrown away. The more complicated options just provide more flexibility for future use cases.

As it happens, the Miner actor has a separate bitfield of all allocated sector numbers so the HAMT check is redundant and will probably be removed (in https://github.com/filecoin-project/specs-actors/issues/1241). But it would be more robust to keep it, if it were free in terms of state blocks accessed.

@ZenGround0 @Stebalien

### PR DESCRIPTION
**Pull request by ZenGround0**
Closes #75 with the `SetIfAbsent` approach

* `modifyValue` now returns a boolean indicating whether or not it has changed the subtree
* `modifyValue` now takes in an overwrite boolean which only allows overwriting existing keys when true
* `SetIfAbsent` exported function

Tests still needed

### CODE REVIEW

File: hamt.go [hunk 1]
```diff
@@ -599,11 +628,12 @@ func (n *Node) modifyValue(ctx context.Context, hv *hashBits, k []byte, v *cbg.D
...
-			return err
+			return false, err
 		}
 
-		if err := chnd.modifyValue(ctx, hv, k, v); err != nil {
-			return err
+		modified, err := chnd.modifyValue(ctx, hv, k, v, overwrite)
+		if err != nil {
+			return false, err
 		}
```
Comment (ZenGround0) [inline hamt.go:637]: Now that we are propagating `modified` I am really tempted to set `child.dirty = modified` to optimize away flushes of duplicate keys.  This would require updating the spec of [v3 hamt draft fip](https://github.com/filecoin-project/FIPs/pull/39) which should be straightforward.  @Stebalien @anorth does this seem worth it?

File: hamt.go [hunk 2]
```diff
@@ -483,7 +484,34 @@ func (n *Node) Set(ctx context.Context, k string, v interface{}) error {
...
 	}
 
-	return n.modifyValue(ctx, &hashBits{b: n.hash(kb)}, kb, d)
+	_, err := n.modifyValue(ctx, &hashBits{b: n.hash(kb)}, kb, d, true)
+	return err
+}
+
+// SetIfAbsent sets key k to value v, where v is has a MarshalCBOR(bytes.Buffer)
+// method to encode it only if k is not already set. Returns true if the value
+// is set, false otherwise.
```
Comment (anorth) [inline hamt.go:493]: "is set" is slightly ambiguous: set before this operation, or set by this operation?

Following [Redis](https://redis.io/commands/setnx), I think true for "changed by this operation" is good.

Following [Java](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#putIfAbsent-K-V-), it could also be helpful for future use cases to return the replaced value, if any. We could probably do this with an out parameter (and skip if it's nil). This would require yet more plumbing through `modifyValue`. @Stebalien how worthwhile would you think this?

**Review by anorth — COMMENTED** [review]
Looking good, needs tests.

The boolean constants now flying around can be a little hard to follow. Consider adding two new boolean type wrappers for `N/OVERWRITE` for the parameter and `UN/MODIFIED` for the return value, so their meaning is transparent wherever they are used.

**Comment by rvagg** [comment]
I had a bit of a go of extending this but pushing the `MarshalCBOR` right down to where the value is set - so it's not wasted if the value isn't set. But `nil` gets in the way of doing that, it'd probably require an additional argument to `modifyValue()` to make it work nicely. That's an option to reduce some code duplication and avoid unnecessary encode cost if the "is not set" case is going to be common enough.

File: hamt.go (hunk 2 above)
Comment (Stebalien) [reply hamt.go:493]: I can't really see any cases where we'd have use for that out parameter, so I'd avoid it unless we really think it's useful.

File: hamt.go (hunk 1 above)
Comment (Stebalien) [reply hamt.go:637]: That seems reasonable but that shouldn't affect any of the _existing_ functionality, only the Set
ifUnset method.

Comment (ZenGround0) [reply hamt.go:637]: @Stebalien ah right this is true as things stand. I was imagining updating the [modify existing](https://github.com/filecoin-project/go-hamt-ipld/pull/77/files#diff-78db9467839255740916ab4674350eccb3cc54256af60665a46725e2b088ebf8R680) code to 

```golang
if overwrite {
   oldVal := p.Value
   p.Value = v
   return oldVal != p.Value, nil
}
``` 

so that a Set of a key with the same value would not register as modified and we could correctly track that this node is not dirty.

Comment (Stebalien) [reply hamt.go:637]: I see. That works for me but it seems like it might be overkill, honestly. From an implementation perspective, updating on set is probably easier. I'd get @austinabell's thoughts on this.

Comment (austinabell) [reply hamt.go:637]: Hmm doesn't seem that difficult to match, only just skimming now but can put a little more thought into it later.

Sounds reasonable to me though, I think I was considering doing this before I found that the reads/writes needed to be matched so exactly.

Just to be clear though, you are saying on set if the value you are putting in is equal to the one that exists, don't change anything, correct? (I'm confused about why you swap the values here in the case of equal)

Comment (ZenGround0) [reply hamt.go:637]: > you are saying on set if the value you are putting in is equal to the one that exists, don't change anything, correct?

yes

> I'm confused about why you swap the values here in the case of equal

This is just a quick example leading to imperfect code. Better
```golang
if overwrite && p.Value != v {
   p.Value = v
   return true, nil
} else {
   return false, nil
}
```

**Review by anorth — COMMENTED** [review]
LGTM after nits

**Review by anorth — APPROVED** [review]
